File: day3/p2.py
import re

def solve_corrupted_memory(memory):
    # Regex for mul(), do(), and don't() instructions
    mul_pattern = r'mul\((\d+),(\d+)\)'
    do_pattern = r'do\(\)'
    dont_pattern = r'don\'?t\(\)'
    
    # Track multiplication state
    mul_enabled = True
    results = []
    
    # Find all instructions in order
    mul_matches = list(re.finditer(mul_pattern, memory))
    do_matches = list(re.finditer(do_pattern, memory))
    dont_matches = list(re.finditer(dont_pattern, memory))
    
    # Combine and sort all matches by their position
    all_matches = sorted(
        [(m.start(), 'mul', m.group(1), m.group(2)) for m in mul_matches] +
        [(m.start(), 'do') for m in do_matches] +
        [(m.start(), 'dont') for m in dont_matches],
        key=lambda x: x[0]
    )
    
    # Process instructions in order
    for match in all_matches:
        if match[1] == 'do':
            mul_enabled = True
        elif match[1] == 'dont':
            mul_enabled = False
        elif match[1] == 'mul':
            if mul_enabled:
                results.append(int(match[2]) * int(match[3]))
    
    # Sum all enabled multiplication results
    total_sum = sum(results)
    
    return total_sum, results

# Read the input file
with open('input_puzzle.txt', 'r') as file:
    corrupted_memory = file.read()

# Solve the puzzle
total, individual_results = solve_corrupted_memory(corrupted_memory)

print(f"Total sum of enabled multiplications: {total}")
print(f"Number of enabled mul() instructions: {len(individual_results)}")
print(f"Individual multiplication results: {individual_results}")

File: day3/p2.js
const fs = require("fs");

// Function to solve the corrupted memory
function solveCorruptedMemory(memory) {
    // Regex patterns for mul(), do(), and don't() instructions
    const mulPattern = /mul\((\d+),(\d+)\)/g;
    const doPattern = /do\(\)/g;
    const dontPattern = /don'?t\(\)/g;

    // Track multiplication state
    let mulEnabled = true;
    const results = [];

    // Find all instructions in order
    const mulMatches = [...memory.matchAll(mulPattern)].map(m => ({
        position: m.index,
        type: "mul",
        values: [parseInt(m[1], 10), parseInt(m[2], 10)],
    }));
    const doMatches = [...memory.matchAll(doPattern)].map(m => ({
        position: m.index,
        type: "do",
    }));
    const dontMatches = [...memory.matchAll(dontPattern)].map(m => ({
        position: m.index,
        type: "dont",
    }));

    // Combine and sort all matches by their position
    const allMatches = [...mulMatches, ...doMatches, ...dontMatches].sort(
        (a, b) => a.position - b.position
    );

    // Process instructions in order
    allMatches.forEach(match => {
        if (match.type === "do") {
            mulEnabled = true;
        } else if (match.type === "dont") {
            mulEnabled = false;
        } else if (match.type === "mul" && mulEnabled) {
            results.push(match.values[0] * match.values[1]);
        }
    });

    // Sum all enabled multiplication results
    const totalSum = results.reduce((sum, val) => sum + val, 0);

    return { totalSum, results };
}

// Read the input file
const filePath = "input_puzzle.txt";
try {
    const corruptedMemory = fs.readFileSync(filePath, "utf8");

    // Solve the puzzle
    const { totalSum, results } = solveCorruptedMemory(corruptedMemory);

    console.log(`Total sum of enabled multiplications: ${totalSum}`);
    console.log(`Number of enabled mul() instructions: ${results.length}`);
    console.log(`Individual multiplication results: ${results}`);
} catch (err) {
    console.error(`Error reading file: ${err.message}`);
}
